### ISSUE TITLE
instance MonadUnliftIO Ctl

### ISSUE THREAD

**Issue by arybczak** [re-xyr/cleff#25]
Can it be written? :thinking:

**Comment by re-xyr** [comment]
No - because you'd need `Ctl a -> IO a`:

```haskell
instance MonadUnliftIO Ctl where
  withRunInIO f = Ctl $ fmap Pure $ f \(Ctl m) -> _what
```

`MonadBaseControl` is possible though:

```haskell
instance MonadBaseControl IO Ctl where
  type StM Ctl a = Result a
  liftBaseWith f = Ctl $ fmap Pure $ f unCtl
  restoreM = Ctl . pure
```

though I'd rather not have it because it certainly would cause all kinds of semantic problems.

**Comment by arybczak** [comment]
> (I assume you mean `speff-same-scope`; the other two branches `speff` and `speff-no-io` are abandoned)
> 
> No - because you'd need `Ctl a -> IO a`:
> 
> ```haskell
> instance MonadUnliftIO Ctl where
>   withRunInIO f = Ctl $ fmap Pure $ f \(Ctl m) -> _what
> ```

Right. That's what I thought :( That makes it unviable for real-world usage to me.

> 
> `MonadBaseControl` is possible though:
> 
> ```haskell
> instance MonadBaseControl IO Ctl where
>   type StM Ctl a = Result a
>   liftBaseWith f = Ctl $ fmap Pure $ f unCtl
>   restoreM = Ctl . pure
> ```
> 
> though I'd rather not have it because it certainly would cause all kinds of semantic problems.

True, if the yield or raise can be lost, we're back to square one.

Ok, thanks for exploring this area. I have a feeling that if/when proper support for delimited continuations in merged into GHC, this shouldn't be a problem anymore since with the low-level support you can manipulate the call stack and effectively bypass the type system, so it should be possible to write `MonadUnliftIO Eff` even when delimited continuations are used underneath.

I'm wondering about this since while I don't personally need the `NonDet` effect, it would be nice to be able to have proper support for it in `effectful` / `cleff`. I'm much less convinced about providing support for coroutines, since:

1. In other languages they are usually treated as very low level primitives and used as implementation detail for other control mechanisms.

2. They (as demostrated by `eff`) unlock a can of worms that makes it very hard to reason what will happen, how and where, while preventing these interactions from being able to wreck your world seems to be problematic.

cc @lexi-lambda - my current assumption is that once your patch for delimited continuations is merged, `effectful` and `cleff` can use it to support `NonDet` somewhat "for free", i.e. without compromising existing guarantees and functionality of these libraries (in particular, the existence of `MonadUnliftIO` instance). The `Coroutine` effect still remains an open question to me.

**Comment by re-xyr** [comment]
Alexis and I discussed the `Coroutine` problem last time. From what I understood, we don't want the resumption to escape the scope of `control` specifically because of this one interaction:

```
interpret[1] { DoThenAbort m -> do { m ; abort () } }
  interpret[2] { Quit -> control \k -> pure k }
    do { DoThenAbort $ do { Quit } }

-->

interpret[1] { DoThenAbort m -> do { m ; abort () } }
  interpret[2] { Quit -> control \k -> pure k }
    do { Quit; abort[1] () }

-->

interpret[1] { DoThenAbort m -> do { m ; abort () } }
  interpret[2] { Quit -> control \k -> pure k }
    do { control[2] \k -> pure k; abort[1] () }

-->

interpret[1] { DoThenAbort m -> do { m ; abort () } }
  pure \x -> interpret[2] { Quit -> control \k -> pure k }
    do { x; abort[1] () }

-->

pure \x -> interpret[2] { Quit -> control \k -> pure k }
  do { x; abort[1] () }
```

at which point you get an `abort` without a pairing `interpret`. the pseudocode is a bit handwavy, but you get the point.

In other words, the interaction of

- You can still control/abort after embedding an `esSend` computation in a handler
- Continuations can escape the scope of `control`

causes this problem, and the removal of any one avoids it. This is all speculation though, I haven't proved it.

**Comment by re-xyr** [comment]
@arybczak Note that IO-native `prompt` and `control` doesn't "just work" in all scenarios; particularly, they won't be able to penetrate `forkIO`s - i.e. this won't work:

```haskell
prompt tag $ do
  forkIO $ do
    ...
    control tag ...

-- GHC.Exts.control0#: no matching prompt in the current continuation
```

...which is a good thing because such behavior can't possibly be well defined, but also a bad thing because we haven't really figured out how to rule them out in an effect system.

One way could be to only allow embedding first-order `IO` computations, but this also rules out "good" higher-order IO functions that do not switch threads, like `catch`. The only other way I can think of now is to slap big warning signs all over the docs.

**Comment by arybczak** [comment]
Ideally I'd like to use `prompt` and `control` just for `NonDet` and leave everything else as-is, so `Error` would still use exceptions underneath.

So this would only be a problem if someone spawns threads inside the `NonDet` effect. IIUC the implementation should throw an exception if `control` cannot find a matching `prompt`, but even if it can't, since in `effectful` `Env` is thread-local, maybe I can keep track of active prompts in `Env` directly and detect this particular case to present an user friendly error. On the other hand, I also don't know how delimited continuations will cooperate with mutable `Env`.

My understanding of delimited continuations is still shallow at best, so this all might not work the way I think it does.
